### PR TITLE
Feature/#5/Add Authorization admin middleware

### PR DESCRIPTION
**Pull request by JDavid550**
# Pull Request

## Type of Change

- [ ] Select proper type of change
    - New feature (non-breaking change which adds functionality)
    - Bug fix (non-breaking change which fixes an issue)
    - Breaking change (fix or feature that would cause existing functionality to not work or work differently than before)

## Branch information

* **Resolves**: NOT AN ISSUE
* **Description** Add a Middleware that validates if the user is admin to getById, Create, Update and Delete as of the req.user population.
* **Architectural Decisions**: NOT APPLICABLE

## Checklist

* [x] Adds new tests
* [x] New and existing test suites succeed
* [x] Linter succeeds
* [ ] Updates corresponding documentation

### CODE REVIEW

File: src/middlewares/AdminAuthorization.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+const AdminAuthorization = ( // eslint-disable-line @typescript-eslint/no-explicit-any
+): IRequestHandler  => 
+  async (req:any, res, next): Promise<any> => {
+
+    const userService = new UserService(); 
+
+    if (req) {
+      Log.info(req.user);
+      console.log(req.user);
```
Comment (herrera-ignacio) [inline src/middlewares/AdminAuthorization.ts]: remove console.logs

File: src/middlewares/AdminAuthorization.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import UserService from "../services/UserService";
+import { UserRole } from "../entities/User";
+import Log from "../utils/Log";
+import HttpException from "../exceptions/HttpException";
+import { IRequestHandler } from "../types/IRequestHandler";
+
+
+
+const AdminAuthorization = ( // eslint-disable-line @typescript-eslint/no-explicit-any
```
Comment (herrera-ignacio) [inline src/middlewares/AdminAuthorization.ts]: We should figure out how to provide a correct type to req

File: src/middlewares/AdminAuthorization.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    const userService = new UserService(); 
+
+    if (req) {
+      Log.info(req.user);
+      console.log(req.user);
+      console.log(req.auth);
+      const user = await userService.getByEmail(req.user.email);
+      if (user.role === UserRole.ADMIN) {
+        Log.info(JSON.stringify(user));
```
Comment (herrera-ignacio) [inline src/middlewares/AdminAuthorization.ts]: We do not want to have those kind of troubleshooting logs being committed to

**Review by herrera-ignacio — DISMISSED** [review]

File: src/middlewares/AuthCheck.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import UserService from "../services/UserService";
+import { IRequestHandler } from "../types/IRequestHandler";
+
+
+
+const AuthCheck = ( // eslint-disable-line @typescript-eslint/no-explicit-any
```
Comment (herrera-ignacio) [inline src/middlewares/AuthCheck.ts]: I prefer favoring classes over functions for this type of functionality. A base class will favor subclassing for reusability.

File: src/middlewares/AuthorizationMiddleware.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+const AuthorizationMiddleware = (
+): IRequestHandler  => 
+  async (req: IAuthorizedRequest, res, next): Promise<User> => {
+
+    // Given 
+    let user;
+    const userService = new UserService(); 
+
+    // Validate
+    if (req.auth && req.auth.sub) {
```
Comment (herrera-ignacio) [inline src/middlewares/AuthorizationMiddleware.ts]: if (req.auth?.sub)

File: src/repositories/UserRepository.ts [hunk 6]
```diff
@@ -13,6 +13,10 @@ const UserRepository = (): Repository <User> => Database.getConnection().getRepo
   findByEmail(email: string): Promise<User> {
     return this.findOne({ where: { email } });
   },
+
+  getByAuth0_id(auth0_id: string): Promise<User> {
```
Comment (herrera-ignacio) [inline src/repositories/UserRepository.ts]: Let's use camelCase and avoid underscores:

* `getByAuth0Id` and `auth0Id`

File: src/services/UserService.ts [hunk 7]
```diff
@@ -55,6 +55,17 @@ export default class UserService implements IService<User>{
     return user;
   }
 
+  /**
+   * Get all the specific user by auth0_id
+   * @param auth0_id
+  */
+
+  public async getByAuth0_id(auth0_id:string): Promise<User>{
```
Comment (herrera-ignacio) [inline src/services/UserService.ts]: camelCase please!

File: src/types/IRepository.ts [hunk 8]
```diff
@@ -3,6 +3,7 @@ import { Repository } from "typeorm";
 type IRepository<T> = Repository<T> & {
   findById?(id: number, options?:any):Promise<T>;
   findByEmail?(email:string):Promise<T>;
+  findByAuth0_id?(auth0_id:string):Promise<T>;
```
Comment (herrera-ignacio) [inline src/types/IRepository.ts]: camelCase! 

We are starting to have too many user-specific methods in this interface which is going to be shared across entities. We should create another type that extends this one such as `IUserRepository` and adds `findByEmail` and `findByAuth0Id`.

File: src/validators/User/UserValidator.ts [hunk 9]
```diff
@@ -17,6 +17,10 @@ export default class UserValidator {
   @IsString()
   @IsOptional()
   public lastName?: string;
+
+  @IsString()
+  @IsOptional()
+  public auth0_id?: string;
```
Comment (herrera-ignacio) [inline src/validators/User/UserValidator.ts]: camelCase

File: tsconfig.json [hunk 10]
```diff
@@ -6,9 +6,12 @@
       "module": "commonjs",
       "esModuleInterop": true,
       "target": "esnext",
-      "typeRoots": ["./node_modules/@types"],
       "noImplicitAny": true,
       "moduleResolution": "node",
+      "typeRoots": [
+        "@types",
```
Comment (herrera-ignacio) [inline tsconfig.json]: you shouldn't need to do this

File: src/middlewares/AuthCheck.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+): IRequestHandler  => 
+  async (req:any, res, next): Promise<any> => {
```
Comment (herrera-ignacio) [inline src/middlewares/AuthCheck.ts]: We should not commit `@typescript-eslint/no-explicit-any` unless there's a very specific reason to do it.

We must avoid dealing with a `req: any` in our code.

File: src/middlewares/AuthorizationMiddleware.ts [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+const AuthorizationMiddleware = (
+): IRequestHandler  => 
+  async (req: IAuthorizedRequest, res, next): Promise<User> => {
+
+    // Given 
+    let user;
+    const userService = new UserService(); 
+
+    // Validate
+    console.log(req.auth);
```
Comment (herrera-ignacio) [inline src/middlewares/AuthorizationMiddleware.ts]: Remove console.log

File: src/middlewares/AuthorizationMiddleware.ts [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    if (req.auth && req.auth.sub) {
+      user = await userService.getByAuth0_id(req.auth.sub);
```
Comment (herrera-ignacio) [inline src/middlewares/AuthorizationMiddleware.ts]: We should not query for a user in an authorization middleware. This is the responsibility of authentication middleware. Authorization should only check whether there's a user in the request or not and if it is then validate role permissions.

File: src/middlewares/AuthorizationMiddleware.test.ts [hunk 14]
```diff
@@ -0,0 +1,43 @@
+import sinon from "sinon";
+import getRequestMock from "../mocks/RequestMock";
+import AdminAuthorization from "./AuthorizationMiddleware";
```
Comment (herrera-ignacio) [inline src/middlewares/AuthorizationMiddleware.test.ts]: We should not name this import as `AdminAuthorization` as it is going to be a base class for us. Instead import it as class name and add test cases for admin users.

File: src/middlewares/AuthCheck.ts [hunk 15]
```diff
@@ -0,0 +1,21 @@
+import UserService from "../services/UserService";
```
Comment (herrera-ignacio) [inline src/middlewares/AuthCheck.ts]: This class seems to not be used anywhere. Furthermore, it seems the logic can be moved to JWTCheckMiddleware.

**Review by herrera-ignacio — CHANGES_REQUESTED** [review]

File: src/middlewares/AuthCheck.ts (hunk 4 above)
Comment (JDavid550) [reply src/middlewares/AuthCheck.ts]: This file was deleted

**Pull request closed without merging**